src/basic/terminal-util.c

```diff
@@ -2436,6 +2452,163 @@ int terminal_get_size_by_dsr(
         return r;
 }
 
+#define MAX_TERMINFO_LENGTH 64
+/* python -c 'print("".join(hex(ord(i))[2:] for i in "TN").upper())' */
+#define DCS_TERMINFO_Q ANSI_DCS "+q" "544E" ANSI_ST
+/* The answer is either 0+r… (invalid) or 1+r… (OK). */
+#define DCS_TERMINFO_R0 ANSI_DCS "0+r" "544E" ANSI_ST
+#define DCS_TERMINFO_R1 ANSI_DCS "1+r" "544E" "=" /* This is followed by Pt ST. */
+
+static int scan_terminfo_response(
+                const char *buf,
+                size_t size,
+                char **ret_name) {
+        int r;
+
+        assert(buf);
+        assert(ret_name);
+
+        /* Check if the have enough space for an answer */
+        if (size < STRLEN(DCS_TERMINFO_R0))
+                return 0;
+
+        /* Check if the terminating sequence is present */
+        if (memcmp(buf + size - STRLEN(ANSI_ST), ANSI_ST, STRLEN(ANSI_ST)) != 0)
+                return 0;
+
+        if (size <= STRLEN(DCS_TERMINFO_R1 ANSI_ST))
+                return -EINVAL;  /* The answer is invalid or empty */
+
+        if (memcmp(buf, DCS_TERMINFO_R1, STRLEN(DCS_TERMINFO_R1)) != 0)
+                return -EINVAL;  /* The answer is not valid */
+
+        _cleanup_free_ void *dec = NULL;
+        size_t dec_size;
+        r = unhexmem_full(buf + STRLEN(DCS_TERMINFO_R1), size - STRLEN(DCS_TERMINFO_R1 ANSI_ST),
+                          /* secure= */ false,
+                          &dec, &dec_size);
+        if (r < 0)
+                return r;
+
+        if (strlen(dec) != dec_size || string_has_cc(dec, NULL))
+                return -EUCLEAN;
+
+        *ret_name = TAKE_PTR(dec);
+        return 1;
+}
+
+int terminal_get_terminfo_by_dcs(
+                int input_fd,
+                int output_fd,
+                char **ret_name) {
+
+        _cleanup_close_ int nonblock_input_fd = -EBADF;
+        int r;
+
+        assert(ret_name);
+
+        r = terminal_verify_same(STDIN_FILENO, STDOUT_FILENO);
+        if (r < 0)
+                return r;
+
+        struct termios old_termios;
+        if (tcgetattr(STDIN_FILENO, &old_termios) < 0)
+                return -errno;
+
+        struct termios new_termios = old_termios;
+        termios_disable_echo(&new_termios);
+
+        if (tcsetattr(STDIN_FILENO, TCSADRAIN, &new_termios) < 0)
+                return -errno;
+
+        r = loop_write(STDOUT_FILENO, DCS_TERMINFO_Q, SIZE_MAX);
+        if (r < 0)
+                goto finish;
+
+        /* Open a 2nd input fd, in non-blocking mode, so that we won't ever hang in read() should someone
+         * else process the POLLIN. */
+
+        nonblock_input_fd = fd_reopen(STDIN_FILENO, O_RDONLY|O_CLOEXEC|O_NONBLOCK|O_NOCTTY);
```

Comment:
Please follow #37501 here too

Reply:
Following Lennart's suggestion, the fd is now opened in non-blocking mode from the beginning, so fd_reopen is not needed and the issue is sidestepped altogether.